src/coreclr/vm/jitinterface.cpp

```diff
@@ -1805,6 +1806,85 @@ uint32_t CEEInfo::getThreadLocalFieldInfo (CORINFO_FIELD_HANDLE  field, bool isG
     return typeIndex;
 }
 
+#ifndef _MSC_VER
+
+void* getThreadStaticDescriptor(uint8_t* p)
+{
+    _ASSERTE_MSG((p[0] == 0x66 && p[1] == 0x48 && p[2] == 0x8d && p[3] == 0x3d),
+        "Unexpected instruction - this can happen when this is not compiled in .so (e.g. for single file)");
+
+    // At this point, `p` contains the instruction pointer and is pointing to the above opcodes.
+    // These opcodes are patched by the dynamic linker.
+    // Move beyond the opcodes that we have already checked above.
+    p += 4;
+
+    // The descriptor address is located at *p at this point. Ready that and add
+    // it to the instruction pointer to locate the address of `ti` that will be used
+    // to pass to __tls_get_addr during execution.
+    return *(uint32_t*)p + (p + 4);
+}
+
+void* getNonGCMaxThreadStaticDescriptor()
+{
+    uint8_t* p;
+    __asm__("leaq 0(%%rip), %%rbx\n"
+            "data16\n"
+            "leaq t_NonGCMaxThreadStaticBlocks@TLSGD(%%rip), %%rdi\n"
+            "data16\n"
+            "data16\n"
+            "rex64\n"
+            "callq __tls_get_addr\n"
+            : "=b"(p));
+
+    return getThreadStaticDescriptor(p);
+}
+
+void* getNonGCThreadStaticBlockDescriptor()
+{
+    uint8_t* p;
+    __asm__("leaq 0(%%rip), %%rbx\n"
+            "data16\n"
+            "leaq t_NonGCThreadStaticBlocks@TLSGD(%%rip), %%rdi\n"
+            "data16\n"
+            "data16\n"
+            "rex64\n"
+            "callq __tls_get_addr\n"
+            : "=b"(p));
+
+    return getThreadStaticDescriptor(p);
+}
+
+void* getGCMaxThreadStaticDescriptor()
+{
+    uint8_t* p;
+    __asm__("leaq 0(%%rip), %%rbx\n"
+            "data16\n"
+            "leaq t_GCMaxThreadStaticBlocks@TLSGD(%%rip), %%rdi\n"
+            "data16\n"
+            "data16\n"
+            "rex64\n"
+            "callq __tls_get_addr\n"
+            : "=b"(p));
+
+    return getThreadStaticDescriptor(p);
+}
+
+void* getGCThreadStaticBlockDescriptor()
+{
+    uint8_t* p;
+    __asm__("leaq 0(%%rip), %%rbx\n"
+            "data16\n"
+            "leaq t_GCThreadStaticBlocks@TLSGD(%%rip), %%rdi\n"
+            "data16\n"
+            "data16\n"
+            "rex64\n"
+            "callq __tls_get_addr\n"
+            : "=b"(p));
+
+    return getThreadStaticDescriptor(p);
+}
+#endif
+
```

Comment:
```suggestion
void* getModuleThreadStaticDescriptor()
{
    uint8_t* p;
    __asm__("leaq 0(%%rip), %%rbx\n"
            "data16\n"
            "leaq t_NonGCMaxThreadStaticBlocks@TLSLDM(%%rip), %%rdi\n"
            "data16\n"
            "data16\n"
            "rex64\n"
            "callq __tls_get_addr\n"
            : "=b"(p));

    return getThreadStaticDescriptor(p);
}
#endif

```

Reply:
Notice that `TLSGD` was changed to `TLSLDM`. It does not matter which thread local you use for it, any thread local from libcoreclr.so would work.

Reply:
I think we will need 2 versions of this method - one for `t_NonGCMaxThreadStaticBlocks` and other for `t_GCMaxThreadStaticBlocks`

Reply:
If we use one descriptor for all `t_` statics, we will be able to better CSE the `_tls_get_addr` call in methods that access both GC and non-GC statics.

If we have two different descriptors for GC and non-GC statics, we will be able to save a few bytes of code size on x64 if the method accesses GC statics only. (I do not think it will save anything on arm64.)

Out of these two options, I think the first one is better tradeoff.

Reply:
> Notice that `TLSGD` was changed to `TLSLDM`.

Interestingly, that crashes the clang.

```   0.    Program arguments: /usr/bin/clang++-10 -DBUILDENV_CHECKED=1 -DDEBUG -DDEBUGGING_SUPPORTED -DDISABLE_CONTRACTS -DEnC_SUPPORTED -DFEATURE_ARRAYSTUB_AS_IL -DFEATURE_BASICFREEZE -DFEATURE_CODE_VERSIONING -DFEATURE_COLLECTIBLE_TYPES -DFEATURE_COMWRAPPERS -DFEATURE_CORECLR -DFEATURE_DBGIPC_TRANSPORT_DI -DFEATURE_DBGIPC_TRANSPORT_VM -DFEATURE_DEFAULT_INTERFACES -DFEATURE_EH_FUNCLETS -DFEATURE_EVENTSOURCE_XPLAT -DFEATURE_EVENT_TRACE -DFEATURE_HIJACK -DFEATURE_ICASTABLE -DFEATURE_INSTANTIATINGSTUB_AS_IL -DFEATURE_ISYM_READER -DFEATURE_MANAGED_ETW -DFEATURE_MANAGED_ETW_CHANNELS -DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES -DFEATURE_MULTICASTSTUB_AS_IL -DFEATURE_MULTICOREJIT -DFEATURE_MULTIREG_RETURN -DFEATURE_ON_STACK_REPLACEMENT -DFEATURE_PAL_ANSI -DFEATURE_PERFMAP -DFEATURE_PERFTRACING -DFEATURE_PGO -DFEATURE_PORTABLE_SHUFFLE_THUNKS -DFEATURE_PROFAPI_ATTACH_DETACH -DFEATURE_READYTORUN -DFEATURE_REJIT -DFEATURE_REMOTE_PROC_MEM -DFEATURE_STANDALONE_GC -DFEATURE_STUBS_AS_IL -DFEATURE_SVR_GC -DFEATURE_SYMDIFF -DFEATURE_TIERED_COMPILATION -DFEATURE_USE_ASM_GC_WRITE_BARRIERS -DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP -DHOST_64BIT -DHOST_AMD64 -DHOST_UNIX -DPROFILING_SUPPORTED -DTARGET_64BIT -DTARGET_AMD64 -DTARGET_LINUX -DTARGET_UNIX -DUNICODE -DUNIX_AMD64_ABI -DUNIX_AMD64_ABI_ITF -DURTBLDENV_FRIENDLY=Checked -DWRITE_BARRIER_CHECK -D_BLD_CLR -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64 -D_SECURE_SCL=0 -D_UNICODE -I/home/kpathak/git/runtime/artifacts/obj/coreclr/linux.x64.Checked/vm/wks -I/home/kpathak/git/runtime/src/coreclr/vm/wks -I/home/kpathak/git/runtime/src/coreclr/vm -I/home/kpathak/git/runtime/src/native -I/home/kpathak/git/runtime/src/coreclr/pal/prebuilt/inc -I/home/kpathak/git/runtime/artifacts/obj -I/home/kpathak/git/runtime/src/coreclr/pal/inc -I/home/kpathak/git/runtime/src/coreclr/pal/inc/rt -I/home/kpathak/git/runtime/src/coreclr/pal/src/safecrt -I/home/kpathak/git/runtime/src/coreclr/inc -I/home/kpathak/git/runtime/src/coreclr/debug/inc -I/home/kpathak/git/runtime/src/coreclr/debug/inc/amd64 -I/home/kpathak/git/runtime/src/coreclr/debug/inc/dump -I/home/kpathak/git/runtime/src/coreclr/md/inc -I/home/kpathak/git/runtime/src/coreclr/classlibnative/bcltype -I/home/kpathak/git/runtime/src/coreclr/classlibnative/cryptography -I/home/kpathak/git/runtime/src/coreclr/classlibnative/inc -I/home/kpathak/git/runtime/artifacts/obj/coreclr/linux.x64.Checked/inc -I/home/kpathak/git/runtime/src/coreclr/hosts/inc -I/home/kpathak/git/runtime/src/coreclr/minipal -I/home/kpathak/git/runtime/src/coreclr/pal/inc/rt/cpp -I/home/kpathak/git/runtime/src/coreclr/nativeresources -I/home/kpathak/git/runtime/src/coreclr/vm/amd64 -I/home/kpathak/git/runtime/src/coreclr/vm/../interop/inc -I/home/kpathak/git/runtime/src/native/libs/System.IO.Compression.Native -I/home/kpathak/git/runtime/src/native/libs/Common -I/home/kpathak/git/runtime/src/coreclr/vm/eventing/eventpipe -fPIC -O2 -g -Wall -Wno-null-conversion -fno-omit-frame-pointer -fms-extensions -fwrapv -fstack-protector-strong -Werror -Wno-unused-variable -Wno-unused-value -Wno-unused-function -Wno-tautological-compare -Wno-unknown-pragmas -Wimplicit-fallthrough -Wno-invalid-offsetof -Wno-unused-but-set-variable -ffp-contract=off -fno-rtti -Wno-unknown-warning-option -ferror-limit=4096 -Wno-unused-private-field -Wno-constant-logical-operand -Wno-pragma-pack -Wno-incompatible-ms-struct -Wno-reserved-identifier -Wno-unsafe-buffer-usage -Wno-single-bit-bitfield-constant-conversion -Wno-cast-function-type-strict -fsigned-char -fvisibility=hidden -ffunction-sections -nostdinc -std=gnu++11 -Xclang -include-pch -Xclang /home/kpathak/git/runtime/artifacts/obj/coreclr/linux.x64.Checked/vm/wks/CMakeFiles/cee_wks_core.dir/cmake_pch.hxx.pch -o CMakeFiles/cee_wks_core.dir/__/jitinterface.cpp.o -c /home/kpathak/git/runtime/src/coreclr/vm/jitinterface.cpp 
  1.    <eof> parser at end of file
  2.    Code generation
  #0 0x00007f561c0094ff llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x9814ff)
  #1 0x00007f561c0077b0 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x97f7b0)
  #2 0x00007f561c008c4d llvm::sys::CleanupOnSignal(unsigned long) (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x980c4d)
  #3 0x00007f561bf5ee60 (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x8d6e60)
  #4 0x00007f56227d8420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
  #5 0x0000000005a4bd18 
  clang: error: clang frontend command failed due to signal (use -v to see invocation)
  clang version 10.0.0-4ubuntu1 
  Target: x86_64-pc-linux-gnu
  Thread model: posix
  InstalledDir: /usr/bin
  clang: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
  clang: note: diagnostic msg: 
  ********************
  
  PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
  Preprocessed source(s) and associated run script(s) are located at:
  clang: note: diagnostic msg: /tmp/jitinterface-26447f.cpp
  clang: note: diagnostic msg: /tmp/jitinterface-26447f.sh
  clang: note: diagnostic msg: 
  
  ********************
```